File: PY/src/core/engine.py
import random
from core.snake import Direction, Snake, _is_opposite

REWARD_APPLE     =  1.0
REWARD_DEATH     = -1.0
REWARD_STEP      =  0.0

_DIRECTION_OFFSETS: dict[Direction, tuple[int, int]] = {
    Direction.UP:    ( 0, -1),
    Direction.DOWN:  ( 0,  1),
    Direction.LEFT:  (-1,  0),
    Direction.RIGHT: ( 1,  0),
}

class Engine:
    def __init__(self):
        self.width = 20
        self.height = 20
        self.reset()

    # ---------------------------------------------
    # Public API
    # ---------------------------------------------
    
    def reset(self) -> dict:
        start = (self.width // 2, self.height // 2)
        self.snake = Snake(start_pos=start)
        self.apple = self._spawn_apple()
        self.score = 0
        self.steps = 0
        self.steps_since_apple = 0
        self.done = False
        return self.get_state()

    def update(self, direction: Direction | None = None) -> tuple[dict, float, bool]:
        if self.done:
            return self.get_state(), 0.0, True

        ate_apple = self._next_head(direction) == self.apple
        self.snake.move(direction, apple=ate_apple)
        
        self.steps += 1
        self.steps_since_apple += 1

        reward = REWARD_STEP

        if ate_apple:
            self.score += 1
            self.steps_since_apple = 0
            self.apple = self._spawn_apple()
            reward += REWARD_APPLE

        if self._is_dead():
            self.done = True
            reward += REWARD_DEATH

        # Prevent agent infinite loops
        if self.steps_since_apple >= self.width * self.height:
            self.done = True

        return self.get_state(), reward, self.done

    def get_state(self):
        return {
            "snake_positions": self.snake.positions,
            "head": self.snake.head.position,
            "direction": self.snake.head.direction,
            "apple": self.apple,
            "score": self.score,
            "steps": self.steps,
            "hunger": self.steps_since_apple,
            "done": self.done,
            "board_size": (self.width, self.height)
        }

    # ---------------------------------------------
    # Helpers
    # ---------------------------------------------

    def _next_head(self, direction: Direction | None):
        if direction is None or _is_opposite(direction, self.snake.head.direction):
            direction = self.snake.head.direction

        x, y = self.snake.head.position
        dx, dy = _DIRECTION_OFFSETS[direction]
        return (x + dx, y + dy)

    def _is_dead(self) -> bool:
        x, y = self.snake.head.position
        if not (0 <= x < self.width and 0 <= y < self.height):
            return True
        return self.snake.head.position in self.snake.positions[1:]

    def _spawn_apple(self) -> tuple[int, int] | None:
        occupied = set(self.snake.positions)
        free_cells = [
            (x, y)
            for x in range(self.width)
            for y in range(self.height)
            if (x, y) not in occupied
        ]
        return random.choice(free_cells) if free_cells else None

File: CPP/core/engine.cpp
#include "engine.h"
#include <cstddef>
#include <optional>
#include <random>
#include <vector>

Engine::Engine() : width(20), height(20), snake({width / 2, height / 2}), rng(std::random_device{}()) {
    reset();
}

State Engine::reset() {
    Position start{width / 2, height / 2};
    snake = Snake(start);

    state = State{};
    state.boardSize = {width, height};
    state.apple = spawnApple().value_or({0, 0});
    return getState();
}

GymState Engine::update(std::optional<Direction> dir) {
    if (state.done) {
        return GymState{
            .state = getState(),
            .reward = 0.0f,
            .done = true
        };
    }

    const Position nHead = nextHead(dir);
    const bool ateApple = (nHead == state.apple);

    Direction actualDir = dir.value_or(snake.getDirection());
    if (isOpposite(actualDir, snake.getDirection())) {
        actualDir = snake.getDirection();
    }

    snake.move(actualDir, ateApple);

    state.steps++;
    state.hunger++;
    float stepReward = REWARD_STEP;

    if (ateApple) {
        state.score++;
        state.hunger = 0;
        state.apple = spawnApple().value_or({0, 0});
        stepReward += REWARD_APPLE;
    }

    if (isDead()) {
        state.done = true;
        stepReward += REWARD_DEATH;
    }

    if (state.hunger >= width * height) state.done = true;

    return GymState{
        .state = getState(),
        .reward = stepReward,
        .done = state.done
    };
}

State Engine::getState() const {
    State current_state = state;
    current_state.snakePositions = snake.getPositions();
    current_state.head = snake.getHead();
    current_state.direction = snake.getDirection();
    return current_state;
}

Position Engine::nextHead(std::optional<Direction> dir) {
    const Direction snakeDir = snake.getDirection();
    Direction finalDir = dir.value_or(snakeDir);

    if (isOpposite(finalDir, snakeDir)) {
        finalDir = snakeDir;
    }

    const Position head = snake.getHead();
    const Position offsets = directionOffset(finalDir);
    return {head.x + offsets.x, head.y + offsets.y};
}

bool Engine::isDead() const {
    const Position head = snake.getHead();
    if (head.x < 0 || head.x >= width ||
        head.y < 0 || head.y >= height)
        return true;

    return snake.getCannibalism();
}

std::optional<Position> Engine::spawnApple() {
    const std::vector<Position>& bodyPositions = snake.getPositions();
    const size_t snakeLen = bodyPositions.size();
    const size_t totalCells = static_cast<size_t>(width * height);

    if (snakeLen >= totalCells) return std::nullopt;

    std::uniform_int_distribution<int> dx(0, width - 1);
    std::uniform_int_distribution<int> dy(0, height - 1);

    if (snakeLen * 2 < totalCells) {
        Position p;
        do { 
            p = {dx(rng), dy(rng)}; 
        } while (std::find(bodyPositions.begin(), bodyPositions.end(), p) != bodyPositions.end());
        return p;
    }

    std::array<bool, 400> mask = {};
    for (const auto& p : bodyPositions)
        mask[static_cast<size_t>(p.y * width + p.x)] = true;

    std::vector<Position> freeCells;
    freeCells.reserve(totalCells - snakeLen);

    for (int y = 0; y < height; ++y)
        for (int x = 0; x < width; ++x)
            if (!mask[static_cast<size_t>(y * width + x)])
                freeCells.push_back({x, y});

    std::uniform_int_distribution<size_t> dist(0, freeCells.size() - 1);
    return freeCells[dist(rng)];
}